### PR TITLE
Feat/cart feature

### PR DESCRIPTION
**Pull request by VirAlex**
## Objective
What: Add Cart Feature
Why: Key feature
How:  With Recoil + Local Storage and added the event sending functionnality to link it with Insights

## Type

- [x] New Feature

## Tested

✅

## Documented

- [x] Have you documented in changed file using comments
- [x] Have you added instructions to the README if needed?

### CODE REVIEW

**Comment by matthew1809** [comment]
Maybe a little animation on the add to cart button to show it's been successful? Nothing happens when you click it, so you're not sure if it worked or not

**Comment by matthew1809** [comment]
You show all sizes for each product in cart, which looks cluttered, maybe just show one?

**Comment by VirAlex** [comment]
> You show all sizes for each product in cart, which looks cluttered, maybe just show one?

The issue is the single product where I click to add in cart doesn't have size in attribute, in attribute it's all mixed. It will necessit to implement an entire logic for size.

**Review by HugoWith — APPROVED** [review]
APPROVED GG 🎉

**Comment by VirAlex** [comment]
> Maybe a little animation on the add to cart button to show it's been successful? Nothing happens when you click it, so you're not sure if it worked or not

All right, nice review 🙏

**Comment by matthew1809** [comment]
Please abstract the cart functionality to take it out of hits and into it's own hook lib, so you can call it from where it's used, without cluttering the code

**Comment by matthew1809** [comment]
There are some unused variables in your functions, for example the cartItem const, can you double check why there are used?

**Comment by matthew1809** [comment]
Especially in useEffects, can you comment on every line that you can to make it clear what the logic is doing and why it's needed? This will also help you see opportunities to refactor.

**Comment by matthew1809** [comment]
> > You show all sizes for each product in cart, which looks cluttered, maybe just show one?
> 
> The issue is the single product where I click to add in cart doesn't have size in attribute, in attribute it's all mixed. It will necessit to implement an entire logic for size.

Would recommend just randomly picking one size to show, doesn't matter which one

**Comment by VirAlex** [comment]
> > > You show all sizes for each product in cart, which looks cluttered, maybe just show one?
> 
> > 
> 
> > The issue is the single product where I click to add in cart doesn't have size in attribute, in attribute it's all mixed. It will necessit to implement an entire logic for size.
> 
> 
> 
> Would recommend just randomly picking one size to show, doesn't matter which one

Good idea 👌🙏

**Comment by VirAlex** [comment]
> Please abstract the cart functionality to take it out of hits and into it's own hook lib, so you can call it from where it's used, without cluttering the code

Maybe a bit tricky because of Recoil `useRecoil.....` can't be called in customHook
[Issue Recoil](https://github.com/facebookexperimental/Recoil/issues/621)

File: src/components/carousels/HitsCarousel.jsx
Comment (VirAlex) [inline src/components/carousels/HitsCarousel.jsx:42]: Add some comments

File: src/components/carousels/HitsCarousel.jsx
Comment (VirAlex) [inline src/components/carousels/HitsCarousel.jsx:45]: Remove the unused const 🙏 @matthew1809

File: src/components/hits/Hits.jsx
Comment (VirAlex) [inline src/components/hits/Hits.jsx:114]: Comments 👍

File: src/components/cart/ArticlesCard.jsx
Comment (VirAlex) [inline src/components/cart/ArticlesCard.jsx:117]: Randomise the size

**Comment by VirAlex** [comment]
> > Please abstract the cart functionality to take it out of hits and into it's own hook lib, so you can call it from where it's used, without cluttering the code
> 
> Maybe a bit tricky because of Recoil `useRecoil.....` can't be called in customHook [Issue Recoil](https://github.com/facebookexperimental/Recoil/issues/621)

@matthew1809 @HugoWith centralisation of the function ✅

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (matthew1809) [inline src/pages/productDetailsPage/ProductDetails.jsx:165]: what happens if the SE isn't using one of these attributes because they don't have the data for example colourHexa?

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (matthew1809) [inline src/pages/productDetailsPage/ProductDetails.jsx:97]: feels like these needs more comments, not exactly clear to me what's going on

File: src/pages/homepage/HomePage.jsx
Comment (matthew1809) [inline src/pages/homepage/HomePage.jsx:45]: Not clear from component name that this is cart, how about naming the component something more specific like CartModal?

File: src/config/cartFunctions.js
Comment (matthew1809) [inline src/config/cartFunctions.js:1]: Is there an atom in featuresConfig to turn on or off the cart? Not everyone will want to show it

File: src/components/header/components/Navigation.jsx
Comment (matthew1809) [inline src/components/header/components/Navigation.jsx:85]: remove commented code

**Review by matthew1809 — CHANGES_REQUESTED** [review]
Have you checked if this can be lazy loaded and how it affects the bundle size? 

Overall, really well done, Alex! Left some small comments, mostly cleanup and naming.

File: src/config/cartFunctions.js
Comment (VirAlex) [reply src/config/cartFunctions.js:1]: I think like we said with @HugoWith it can be always active it still work because it's not linked to an attribute or an index,...

File: src/utils/clampCalcFunction.js
Comment (VirAlex) [reply src/utils/clampCalcFunction.js:10]: Cleaned with more recent branch

File: src/scss/partials/_mixin.scss
Comment (VirAlex) [reply src/scss/partials/_mixin.scss:305]: Cleaned with more recent branch

File: src/scss/partials/_globals.scss
Comment (VirAlex) [reply src/scss/partials/_globals.scss:18]: Cleaned with more recent branch

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (VirAlex) [reply src/pages/productDetailsPage/ProductDetails.jsx]: remove because it was old function it's all centralised in a selector now in recoil

File: src/pages/homepage/HomePage.jsx
Comment (VirAlex) [reply src/pages/homepage/HomePage.jsx:45]: ✅

Comment (VirAlex) [reply src/pages/homepage/HomePage.jsx]: https://github.com/algolia/algolia-react-boilerplate/pull/436#:~:text=now%20in%20recoil-,fix%3A%20rename%20modal%20in%20Cart%20Modal,-c69041e

File: src/components/header/components/Navigation.jsx
Comment (VirAlex) [reply src/components/header/components/Navigation.jsx:85]: Cleaned with more recent branch

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (VirAlex) [reply src/pages/productDetailsPage/ProductDetails.jsx]: > Have you checked if this can be lazy loaded and how it affects the bundle size?
> 
> Overall, really well done, Alex! Left some small comments, mostly cleanup and naming.

No I didn't realize that

Comment (VirAlex) [reply src/pages/productDetailsPage/ProductDetails.jsx]: You can watch centralised function : 
`src/config/cartFunctions.js`

File: src/config/cartFunctions.js
Comment (HugoWith) [reply src/config/cartFunctions.js:1]: My Bad Alex, I had a poor judgement on this, I thought it won't be a problem for SE to have it active. Let's create an Atom then, I can do it, no worries 
@VirAlex I think we should delete the PR as the branch is not fully ready to be merged. I'd like to work on some issues as well. What do you think ?

Comment (VirAlex) [reply src/config/cartFunctions.js:1]: I think we should create an atom yes but keep open this PR to have the history of it, actually all @matthew1809 have been fixed because since this comment we continued to work on it what's missing actually is the lazy investigation @HugoWith

**Comment by VirAlex** [comment]
> Have you checked if this can be lazy loaded and how it affects the bundle size?
> 
> Overall, really well done, Alex! Left some small comments, mostly cleanup and naming.

I think if I refer to this SO link like it's called on all pages it's not necessary to use `Lazy`
[Link](https://stackoverflow.com/a/60396594/12073633)

File: src/config/cartFunctions.js
Comment (VirAlex) [reply src/config/cartFunctions.js:1]: ✅

Comment (HugoWith) [reply src/config/cartFunctions.js:1]: Thanks a lot @VirAlex 🙏

**Comment by VirAlex** [comment]
> Have you checked if this can be lazy loaded and how it affects the bundle size?
> 
> Overall, really well done, Alex! Left some small comments, mostly cleanup and naming.

Done the lazy loading ✅
https://github.com/algolia/algolia-react-boilerplate/pull/436/commits/b8952b45d4b38966f289ae1cf2b16c30b97b0dfa

File: src/config/cartFunctions.js
Comment (VirAlex) [reply src/config/cartFunctions.js:1]: https://github.com/algolia/algolia-react-boilerplate/pull/436/commits/db5d9dac04855b9ff13a239caa64cf8934a3b2b1
Done ✅ @matthew1809 @HugoWith

**Comment by VirAlex** [comment]
@matthew1809 normally it's all corrected if you can check if all your requests a done ✅ after that I think we'll can merge it

File: src/utils/clampCalcFunction.js
Comment (matthew1809) [inline src/utils/clampCalcFunction.js:11]: Why comment and not remove?

File: src/utils/clampCalcFunction.js
Comment (matthew1809) [inline src/utils/clampCalcFunction.js:12]: why not return?

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (matthew1809) [reply src/pages/productDetailsPage/ProductDetails.jsx:147]: @VirAlex

File: src/components/header/components/Navigation.jsx
Comment (matthew1809) [inline src/components/header/components/Navigation.jsx:87]: Still seeing a comment here, why leave it in?

**Review by matthew1809 — CHANGES_REQUESTED** [review]

File: src/utils/clampCalcFunction.js
Comment (VirAlex) [reply src/utils/clampCalcFunction.js:12]: ✅

File: src/pages/productDetailsPage/ProductDetails.jsx
Comment (VirAlex) [reply src/pages/productDetailsPage/ProductDetails.jsx:147]: There is already a const `hexaCode` line 158 to check is there is hexacode or not

File: src/components/header/components/Navigation.jsx
Comment (VirAlex) [reply src/components/header/components/Navigation.jsx]: ✅

File: src/utils/clampCalcFunction.js
Comment (VirAlex) [reply src/utils/clampCalcFunction.js]: ✅

**Comment by HugoWith** [comment]
> @matthew1809 normally it's all corrected if you can check if all your requests a done ✅ after that I think we'll can merge it @HugoWith

Once @matthew1809 has given his go, I'll merge and check if it's all good on develop 👌. Thanks @VirAlex

**Review by matthew1809 — APPROVED** [review]
GG!

**Pull request merged**
